pkg/ccl/crosscluster/logical/logical_replication_job.go

```diff
@@ -172,130 +173,263 @@ func (r *logicalReplicationResumer) ingest(
 		req.TableIDs = append(req.TableIDs, pair.SrcDescriptorID)
 	}
 
-	plan, err := client.PlanLogicalReplication(ctx, req)
+	planner, err := makeLogicalReplicationPlannerAndInitialize(
+		ctx,
+		req,
+		jobExecCtx,
+		client,
+		progress,
+		payload,
+		jobID,
+		replicatedTimeAtStart)
+
 	if err != nil {
 		return err
 	}
 
-	dstToSrcDescMap := make(map[int32]descpb.TableDescriptor)
-	for _, pair := range payload.ReplicationPairs {
-		dstToSrcDescMap[pair.DstDescriptorID] = plan.DescriptorMap[pair.SrcDescriptorID]
-	}
+	replanOracle := sql.ReplanOnCustomFunc(
+		measurePlanChange,
+		func() float64 {
+			return crosscluster.ReplanThreshold.Get(jobExecCtx.ExecCfg().SV())
+		},
+	)
 
-	frontier, err := span.MakeFrontierAt(replicatedTimeAtStart, plan.SourceSpans...)
-	if err != nil {
-		return err
-	}
-	for _, resolvedSpan := range progress.Checkpoint.ResolvedSpans {
-		if _, err := frontier.Forward(resolvedSpan.Span, resolvedSpan.Timestamp); err != nil {
-			return err
+	replanner, stopReplanner := sql.PhysicalPlanChangeChecker(ctx,
+		planner.initialPlan,
+		planner.generatePlan,
+		jobExecCtx,
+		replanOracle,
+		func() time.Duration { return crosscluster.ReplanFrequency.Get(jobExecCtx.ExecCfg().SV()) },
+	)
+	metrics := execCfg.JobRegistry.MetricsStruct().JobSpecificMetrics[jobspb.TypeLogicalReplication].(*Metrics)
+
+	execPlan := func(ctx context.Context) error {
+		heartbeatSender := streamclient.NewHeartbeatSender(ctx, client, streampb.StreamID(streamID),
+			func() time.Duration {
+				return heartbeatFrequency.Get(&execCfg.Settings.SV)
+			})
+		defer func() {
+			_ = heartbeatSender.Stop()
+			stopReplanner()
+		}()
```

Comment:
I think rather than making the heartbeat sender in here, you'll want to make it outside of the func so that then you can have another goroutine that you pass to GoAndWait which will do something like:

```
	return heartbeatSender.Wait()
```

But this can happen in a different commit or PR.

Reply:
This I'll handle in a different PR